custom/enikshay/expressions.py

```diff
@@ -119,6 +121,113 @@ def concatenate_strings_expression(spec, context):
     return wrapped
 
 
+class ReferralExpressionBase(JsonObject):
+    person_id_expression = DefaultProperty(required=True)
+
+    def configure(self, person_id_expression):
+        self._person_id_expression = person_id_expression
+
+    def __call__(self, item, context=None):
+        person_id = self._person_id_expression(item, context)
+        domain = context.root_doc['domain']
+        if not person_id:
+            return None
+
+        referral = get_open_referral_case_from_person(domain, person_id)
+        if referral:
+            return self._handle_referral_case(referral)
+        trail = get_latest_trail_case_from_person(domain, person_id)
+        if trail:
+            return self._handle_trail_case(trail, domain)
+        return None
+
+    def _handle_referral_case(self, referral):
+        raise NotImplementedError
+
+    def _handle_trail_case(self, trail, domain):
+        raise NotImplementedError
+
+
+class ReferredBy(ReferralExpressionBase):
+    type = TypeProperty("referred_by")
```

Comment:
That's not currently the pattern we've been using, but I imagine at some point we will want some sort of namespacing. I guess it can't hurt to change now.